source/plugins/go/src/ingestion_token_utils.go

```diff
@@ -0,0 +1,437 @@
+			return imdsAccessToken, 0, err
+		}
+
+	} else {
+		Log("Info Reading IMDS Access Token from file : %s", IMDSTokenPathForWindows)
+		if _, err = os.Stat(IMDSTokenPathForWindows); os.IsNotExist(err) {
+			Log("getAccessTokenFromIMDS: IMDS token file doesnt exist: %s", err.Error())
+			return imdsAccessToken, 0, err
+		}
+		//adding retries incase if we ended up reading the token file while the token file being written
+		for retryCount := 0; retryCount < 3; retryCount++ {
+			responseBytes, err = ioutil.ReadFile(IMDSTokenPathForWindows)
+			if err != nil {
+				Log("getAccessTokenFromIMDS: Could not read IMDS token from file: %s, retryCount: %d", err.Error(), retryCount)
+				time.Sleep(100 * time.Millisecond)
+				continue
+			}
+			break
+	    }
+    }
+
+	// Unmarshall response body into struct
+	var imdsResponse IMDSResponse
+	err = json.Unmarshal(responseBytes, &imdsResponse)
+	if err != nil {
+		Log("getAccessTokenFromIMDS: Error unmarshalling the response: %s", err.Error())
+		return imdsAccessToken, 0, err
+	}
+	imdsAccessToken = imdsResponse.AccessToken
+
+	expiration, err := strconv.ParseInt(imdsResponse.ExpiresOn, 10, 64)
+	if err != nil {
+		Log("Error parsing ExpiresOn field from IMDS response: %s", err.Error())
+		return imdsAccessToken, 0, err
+	}
+	Log("Info getAccessTokenFromIMDS: end")
+	return imdsAccessToken, expiration, nil
+}
+
+func getAgentConfiguration(imdsAccessToken string) (configurationId string, channelId string, err error) {
+	Log("Info getAgentConfiguration: start")
+	configurationId = ""
+	channelId = ""
+	apiVersion := "2020-08-01-preview"
+	var amcs_endpoint *url.URL
+	osType := os.Getenv("OS_TYPE")
+	resourceId := os.Getenv("AKS_RESOURCE_ID")
+	resourceRegion := os.Getenv("AKS_REGION")
+	mcsEndpoint := os.Getenv("MCS_ENDPOINT")
+	amcs_endpoint_string := fmt.Sprintf("https://%s.handler.control.%s%s/agentConfigurations?platform=%s&api-version=%s", resourceRegion, mcsEndpoint, resourceId, osType, apiVersion)
+	amcs_endpoint, err = url.Parse(amcs_endpoint_string)
+
+	var bearer = "Bearer " + imdsAccessToken
+	// Create a new request using http
+	req, err := http.NewRequest("GET", amcs_endpoint.String(), nil)
+	if err != nil {
+		message := fmt.Sprintf("Error creating HTTP request for AMCS endpoint: %s", err.Error())
+		Log(message)
+		return configurationId, channelId, err
+	}
+	req.Header.Set("Authorization", bearer)
+
+	var resp *http.Response = nil
+
+	for i := 0; i < 3; i++ {
+		resp, err = HTTPClient.Do(req)
+		if err != nil {
+			message := fmt.Sprintf("Error calling AMCS endpoint: %s", err.Error())
+			Log(message)
+			SendException(message)
+			continue
+		}
+		if resp != nil && resp.Body != nil {
+			defer resp.Body.Close()
+	    }
+		Log("getAgentConfiguration Response Status: %d", resp.StatusCode)
+		if resp.StatusCode != 200 {
+			message := fmt.Sprintf("getAgentConfiguration Request failed with an error code : %d", resp.StatusCode)
```

Comment:
Same here. We will end up retrying for all non 200s including client errors.

Reply:
good catch. Fixed .